src/EFCore.PG/Query/Internal/NpgsqlQueryableMethodTranslatingExpressionVisitor.cs

```diff
@@ -0,0 +1,56 @@
+using System.Diagnostics.CodeAnalysis;
+
+namespace Npgsql.EntityFrameworkCore.PostgreSQL.Query.Internal;
+
+public class NpgsqlQueryableMethodTranslatingExpressionVisitor : RelationalQueryableMethodTranslatingExpressionVisitor
+{
+    public NpgsqlQueryableMethodTranslatingExpressionVisitor(
+        QueryableMethodTranslatingExpressionVisitorDependencies dependencies,
+        RelationalQueryableMethodTranslatingExpressionVisitorDependencies relationalDependencies,
+        QueryCompilationContext queryCompilationContext)
+        : base(dependencies, relationalDependencies, queryCompilationContext)
+    {
+    }
+
+    protected override bool IsValidSelectExpressionForBulkDelete(
+        SelectExpression selectExpression,
+        EntityShaperExpression entityShaperExpression,
+        [NotNullWhen(true)] out TableExpression? tableExpression)
+    {
+        // The default relational behavior is to allow only single-table expressions, and the only permitted feature is a predicate.
+        // Here we extend this to also inner joins to tables, which we generate via the PostgreSQL-specific USING construct.
+        if (selectExpression.Offset == null
+            && selectExpression.Limit == null
+            && (!selectExpression.IsDistinct || entityShaperExpression.EntityType.FindPrimaryKey() != null)
+            && selectExpression.GroupBy.Count == 0
+            && selectExpression.Having == null
+            && selectExpression.Orderings.Count == 0)
+        {
+            TableExpressionBase? table = null;
+            if (selectExpression.Tables.Count == 1)
+            {
+                table = selectExpression.Tables[0];
+            }
+            else if (selectExpression.Tables.All(t => t is TableExpression or InnerJoinExpression { Table: TableExpression }))
```

Comment:
@smitpatel note this filter to only accept inner joins over tables (since that's what PG USING does/supports).

Reply:
Little sad that only inner join is supported.

Reply:
Yeah, I agree it's sad... Good to add some more tests (multiple inner joins, left/cross...).

I'll take a look at whether USING allows cross apply (lateral join as PG calls it). My gut says no.

Reply:
@smitpatel unfortunately cross apply (lateral join) doesn't work with DELETE's USING:

```sql
DELETE FROM "Order Details" AS o
USING (
    SELECT o0."OrderID", o0."CustomerID", o0."EmployeeID", o0."OrderDate"
    FROM "Orders" AS o0
    WHERE o0."OrderID" < o."OrderID"
    ORDER BY o0."OrderID" NULLS FIRST
    LIMIT 100 OFFSET 0
) AS t
WHERE o."OrderID" < 10276
```

This errors with:

```
42P01: invalid reference to FROM-clause entry for table "o"
```

Note that it is apparently possible to do it with UPDATE's FROM, [as here](https://stackoverflow.com/questions/67221141/syntax-error-lateral-join-with-delete-postgresql):

```sql
UPDATE x
SET foo=bar
FROM a, LATERAL (...)
```

I tried putting LATERAL betwen the subquery and USING in the DELETE above and that didn't work either :cry:

Reply:
Does it work in id in subquery form?

Reply:
You mean the default behavior with the PK comparison? Sure, [here's your test which passes](https://github.com/npgsql/efcore.pg/pull/2449/files#diff-81845b9be5b02109f69bc2211616c63ad71dada661bdea5bfece515a53db0f70R470). It's just that DELETE"s USING clause is apparently only for inner joins...